File: transfer_manager.py
import socket
import threading
import os

TRANSFER_PORT = 54321
BUFFER_SIZE   = 65536

class TransferManager:

    def __init__(self):
        self.on_file_received   = None
        self.on_progress_update = None

    def send_file(self, host_ip: str, filepath: str):
        def _send():
            filesize = os.path.getsize(filepath)
            filename = os.path.basename(filepath)
            try:
                with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                    s.settimeout(10)
                    s.connect((host_ip, TRANSFER_PORT))
                    header = f"{filename}|{filesize}".encode()
                    s.sendall(len(header).to_bytes(4, "big") + header)
                    sent = 0
                    with open(filepath, "rb") as f:
                        while chunk := f.read(BUFFER_SIZE):
                            s.sendall(chunk)
                            sent += len(chunk)
                            if self.on_progress_update:
                                self.on_progress_update(int(sent / filesize * 100))
            except Exception as e:
                print(f"Send error: {e}")
        threading.Thread(target=_send, daemon=True).start()

    def start_receiving(self, save_dir: str = "."):
        def _listen():
            try:
                if not os.path.exists(save_dir):
                    os.makedirs(save_dir)
                with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server:
                    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                    server.bind(("0.0.0.0", TRANSFER_PORT))
                    server.listen(1)
                    conn, _ = server.accept()
                    with conn:
                        header_len = int.from_bytes(conn.recv(4), "big")
                        header     = conn.recv(header_len).decode()
                        filename, filesize = header.split("|")
                        filesize   = int(filesize)
                        save_path  = os.path.join(save_dir, filename)
                        received   = 0
                        with open(save_path, "wb") as f:
                            while received < filesize:
                                chunk = conn.recv(min(BUFFER_SIZE, filesize - received))
                                if not chunk:
                                    break
                                f.write(chunk)
                                received += len(chunk)
                                if self.on_progress_update:
                                    self.on_progress_update(int(received / filesize * 100))
                        if self.on_file_received:
                            self.on_file_received(save_path)
            except Exception as e:
                print(f"Receive error: {e}")
        threading.Thread(target=_listen, daemon=True).start()

    def get_my_ip(self) -> str:
        try:
            s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
            s.connect(("8.8.8.8", 80))
            ip = s.getsockname()[0]
            s.close()
            return ip
        except Exception:
            return "127.0.0.1"

File: AirSyncAndroid/app/src/main/java/org/airsync/TransferManager.java
package org.airsync;

import android.os.Handler;
import android.os.Looper;
import java.io.*;
import java.net.*;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TransferManager {
    private static final int TRANSFER_PORT = 54321;
    private static final int BUFFER_SIZE = 65536;
    private final ExecutorService executor = Executors.newSingleThreadExecutor();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private TransferListener listener;

    public interface TransferListener {
        void onFileReceived(String path);
        void onProgressUpdate(int percent);
    }

    public void setListener(TransferListener listener) {
        this.listener = listener;
    }

    public void sendFile(String hostIp, String filePath) {
        executor.execute(() -> {
            File file = new File(filePath);
            long fileSize = file.length();
            String fileName = file.getName();

            try (Socket socket = new Socket()) {
                socket.connect(new InetSocketAddress(hostIp, TRANSFER_PORT), 10000);
                DataOutputStream dos = new DataOutputStream(socket.getOutputStream());

                String header = fileName + "|" + fileSize;
                byte[] headerBytes = header.getBytes();
                dos.writeInt(headerBytes.length);
                dos.write(headerBytes);

                FileInputStream fis = new FileInputStream(file);
                byte[] buffer = new byte[BUFFER_SIZE];
                int read;
                long sent = 0;
                while ((read = fis.read(buffer)) != -1) {
                    dos.write(buffer, 0, read);
                    sent += read;
                    int progress = (int) ((sent * 100) / fileSize);
                    mainHandler.post(() -> listener.onProgressUpdate(progress));
                }
                dos.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public void startReceiving(String saveDir) {
        executor.execute(() -> {
            try (ServerSocket serverSocket = new ServerSocket(TRANSFER_PORT)) {
                Socket clientSocket = serverSocket.accept();
                DataInputStream dis = new DataInputStream(clientSocket.getInputStream());

                int headerLen = dis.readInt();
                byte[] headerBytes = new byte[headerLen];
                dis.readFully(headerBytes);
                String header = new String(headerBytes);
                String[] parts = header.split("\\|");
                String fileName = parts[0];
                long fileSize = Long.parseLong(parts[1]);

                File file = new File(saveDir, fileName);
                try (FileOutputStream fos = new FileOutputStream(file)) {
                    byte[] buffer = new byte[BUFFER_SIZE];
                    int read;
                    long received = 0;
                    while (received < fileSize && (read = dis.read(buffer, 0, (int) Math.min(buffer.length, fileSize - received))) != -1) {
                        fos.write(buffer, 0, read);
                        received += read;
                        int progress = (int) ((received * 100) / fileSize);
                        mainHandler.post(() -> listener.onProgressUpdate(progress));
                    }
                }
                mainHandler.post(() -> listener.onFileReceived(file.getAbsolutePath()));
            } catch (IOException e) {
                e.printStackTrace();
            }
        });
    }

    public String getMyIp() {
        try (DatagramSocket socket = new DatagramSocket()) {
            socket.connect(InetAddress.getByName("8.8.8.8"), 80);
            return socket.getLocalAddress().getHostAddress();
        } catch (Exception e) {
            return "127.0.0.1";
        }
    }
}
